tests/integration/test_charm.py

```diff
@@ -0,0 +1,161 @@
+    lightkube_client = lightkube.Client(field_manager=CHARM_NAME)
+    k8s_resource_handler = KubernetesResourceHandler(
+        field_manager=CHARM_NAME, template_files=template_files, context={}
+    )
+    load_in_cluster_generic_resources(lightkube_client)
+    k8s_resource_handler.apply()
+
+
+@pytest.fixture(scope="session")
+def namespace(lightkube_client: lightkube.Client):
+    yaml_text = _safe_load_file_to_text(NAMESPACE_FILE)
+    yaml_rendered = yaml.safe_load(yaml_text)
+    for label in TESTING_LABELS:
+        yaml_rendered["metadata"]["labels"][label] = "true"
+    obj = codecs.from_dict(yaml_rendered)
+    lightkube_client.apply(obj)
+
+    yield obj.metadata.name
+
+    delete_all_from_yaml(yaml_text, lightkube_client)
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Build the ngc-integrator charm and deploy it together with related charms."""
+
+    deploy_k8s_resources([PODDEFAULTS_CRD_TEMPLATE])
+
+    # Build and deploy charm from local source folder
+    built_charm_path = await ops_test.build_charm(".")
+
+    # Deploy the charm and wait for active/idle status
+    await ops_test.model.deploy(
+        entity_url=built_charm_path,
+        application_name=CHARM_NAME,
+        trust=True,
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[CHARM_NAME], status="active", raise_on_blocked=True, timeout=300
+    )
+
+    # Deploy resource-dispatcher charm and related charms
+    await ops_test.model.deploy(
+        entity_url=METACONTROLLER_CHARM_NAME,
+        channel="latest/edge",
+        trust=True,
+    )
+    await ops_test.model.wait_for_idle(
+        apps=[METACONTROLLER_CHARM_NAME],
+        status="active",
+        raise_on_blocked=False,
+        raise_on_error=False,
+        timeout=120,
+    )
+
+    await ops_test.model.deploy(
+        entity_url=RESOURCE_DISPATCHER_CHARM_NAME,
+        channel="latest/edge",
+        trust=True,
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[RESOURCE_DISPATCHER_CHARM_NAME],
+        status="active",
+        raise_on_blocked=False,
+        raise_on_error=False,
+        timeout=1200,
+        idle_period=60,
+    )
+
+    await ops_test.model.relate(
+        f"{CHARM_NAME}:pod-defaults", f"{RESOURCE_DISPATCHER_CHARM_NAME}:pod-defaults"
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[CHARM_NAME, RESOURCE_DISPATCHER_CHARM_NAME],
+        status="active",
+        raise_on_blocked=False,
+        raise_on_error=False,
+        timeout=300,
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_new_user_namespace_has_poddefault(
+    ops_test: OpsTest, lightkube_client: lightkube.Client, namespace: str
+):
+    """Test that the Kubeflow user namespace has the PodDefault object."""
+    time.sleep(30)  # sync can take up to 10 seconds for reconciliation loop to trigger
+
+    pod_default = lightkube_client.get(PodDefault, PODDEFAULT_NAME, namespace=namespace)
```

Comment:
It's a nice to have, you could add a context manager here to ensure this call won't raise and then you can do some more assertions afterwards.